web/src/main/java/org/springframework/security/web/access/expression/WebExpressionAuthorizationManager.java

```diff
@@ -0,0 +1,98 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.web.access.expression;
+
+import java.util.function.Supplier;
+
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.expression.EvaluationContext;
+import org.springframework.expression.Expression;
+import org.springframework.expression.ExpressionParser;
+import org.springframework.expression.ParseException;
+import org.springframework.security.access.expression.ExpressionUtils;
+import org.springframework.security.access.expression.SecurityExpressionHandler;
+import org.springframework.security.authorization.AuthorizationDecision;
+import org.springframework.security.authorization.AuthorizationManager;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.web.access.intercept.RequestAuthorizationContext;
+import org.springframework.util.Assert;
+
+/**
+ * An expression-based {@link AuthorizationManager} that determines the access by
+ * evaluating the provided expression.
+ *
+ * @author Evgeniy Cheban
+ * @since 5.7
+ */
+public final class WebExpressionAuthorizationManager
+		implements AuthorizationManager<RequestAuthorizationContext>, InitializingBean {
+
+	private SecurityExpressionHandler<RequestAuthorizationContext> expressionHandler = new DefaultHttpSecurityExpressionHandler();
+
+	private Expression expression;
+
+	private final String expressionString;
+
+	/**
+	 * Creates an instance.
+	 * @param expressionString the raw expression string to parse
+	 */
+	public WebExpressionAuthorizationManager(String expressionString) {
+		Assert.hasText(expressionString, "expressionString cannot be empty");
+		this.expressionString = expressionString;
+	}
+
+	/**
+	 * Sets the {@link SecurityExpressionHandler} to be used. The default is
+	 * {@link DefaultHttpSecurityExpressionHandler}.
+	 * @param expressionHandler the {@link SecurityExpressionHandler} to use
+	 */
+	public void setExpressionHandler(SecurityExpressionHandler<RequestAuthorizationContext> expressionHandler) {
+		Assert.notNull(expressionHandler, "expressionHandler cannot be null");
+		this.expressionHandler = expressionHandler;
+	}
+
+	/**
+	 * Determines the access by evaluating the provided expression.
+	 * @param authentication the {@link Supplier} of the {@link Authentication} to check
+	 * @param context the {@link RequestAuthorizationContext} to check
+	 * @return an {@link AuthorizationDecision} based on the evaluated expression
+	 */
+	@Override
+	public AuthorizationDecision check(Supplier<Authentication> authentication, RequestAuthorizationContext context) {
+		EvaluationContext ctx = this.expressionHandler.createEvaluationContext(authentication.get(), context);
```

Comment:
I wonder if there is something we can do here to continue a delayed lookup of the `Authentication` rather than eagerly obtaining it. Ideally an expression of `permitAll` would not cause the `Authentication` to be accessed so that it does not require the `HttpSession` to be accessed unnecessarily which can be costly for requests to static resources when the `HttpSession` is persisted in an external store like `Redis`.

I think perhaps in another PR we should be made to make this possible. First it would add another constructor to `WebExpressionRoot` that takes a `Supplier<Authentication>`. Next a default method to `SecurityExpressionHandler` that is the same as  createEvaluationContext but accepts a `Supplier<Authentication`. Each implementation would override the default implementation to pass the `Supplier<Authentication>` into the new ExpressionRoot. It would deprecate the non `Supplier<Authentication>` method and constructor.

Reply:
I agree. Also we can use the same approach for Pre/Post `AuthorizationManager`.
I'd like to work on this in another PR.

Reply:
I believe there is a ticket for this: https://github.com/spring-projects/spring-security/issues/9667